### PR TITLE
fix 21905 - case of IFTI failure when combining `ref`, and `alias this` on a static instance

### PR DESCRIPTION
**Pull request by ghost**
This follows #12157. If the chain of alias this ends up on the original type, keep that type because it can be a static instance but still break because of the infinite loop.

### CODE REVIEW

File: test/compilable/issue21905.d [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+{
+    StaticIterable b;
+    alias b this;
+}
+
+struct StaticIterable
+{
+    static Conv b;
+    alias b this;
+}
```
Comment (thewilsonator) [inline test/compilable/issue21905.d:13]: How does this not result in  recursive type error?

File: src/dmd/mtype.d [hunk 2]
```diff
@@ -4929,9 +4929,11 @@ extern (C++) final class TypeFunction : TypeNext
                         Type tat = tab.aliasthisOf();
                         if (!tat || !tat.implicitConvTo(tprm))
                             break;
-                        if (tat == tab || tat == firsttab)
+                        if (tat == tab)
```
[the remark this answers is not in the document]
Comment (RazvanN7) [reply src/dmd/mtype.d:4932]: For example, my assumption is that at this point [1], `m` should be a perfect match. The fact that we end up checking the alias this is problematic. Maybe we can do the transitive alias this check only if `m` is no match.

[1] https://github.com/dlang/dmd/pull/12503/files#diff-e77707f4e2e026ebef4e2d37d5a66da0c617b09e0b8466eb6341804c65bb0b01L4855

**Comment by RazvanN7** [comment]
ping @TungstenHeart

File: src/dmd/mtype.d (hunk 2 above)
[the remark this answers is not in the document]
Comment (TungstenHeart) [reply src/dmd/mtype.d:4932]: indeed, alias this should not be tried.

**Comment by TungstenHeart** [comment]
@RazvanN7, thanks for the help, updated with a better fix.

File: src/dmd/mtype.d (hunk 2 above)
[the remark this answers is not in the document]
Comment (RazvanN7) [reply src/dmd/mtype.d]: I still don't understand how the code gets this far. Why is it not returning when it sees that a `StaticIterable` is passed to a `StaticIterable`?

[the remark this answers is not in the document]
Comment (RazvanN7) [reply src/dmd/mtype.d]: ping @TungstenHeart

**Comment by TungstenHeart** [comment]
ping

**Review by thewilsonator — APPROVED** [review]

**Comment by RazvanN7** [comment]
I am not convinced this is the right fix. I will reiterate my comment, that was ignored:

"I still don't understand how the code gets this far. Why is it not returning when it sees that a StaticIterable is passed to a StaticIterable?"

**Comment by TungstenHeart** [comment]
Sorry, I thought it was more or less approve after the last forced push. You can revert if you like.

**Pull request merged**